### PR TITLE
Add polybft consensus support to edge

### PR DESCRIPTION
**Pull request by epikichi**
# Description
* Adding PolyBFT Consensus option to Docker compose
* Adding `--docker` modes to the `scripts/cluster` script
* Adding `README` to `docker` and `scripts` folder
* Changed `log-level` to `debug` for all containers

## How to use
* `scripts/cluster ibft --docker`  - run docker environment with `ibft` consensus
* `scripts/cluster polybft --docker` run docker environment with `polybft` consensus
* `scripts/cluster (ibft or polybft) --docker destroy` - destroy docker environment (volumes will also be deleted)
* `scripts/cluster (ibft or polybft) --docker stop` - stop docker environment

### CODE REVIEW

File: Makefile
Comment (vcastellm) [inline Makefile]: IMO Makefile is not the place to manage the lifecycle of local env, it's not building any aspect of the binary, this belongs to a `docker` script in `scripts` folder

**Review by vcastellm — CHANGES_REQUESTED** [review]

File: docker/README.md
Comment (vcastellm) [inline docker/README.md]: Even though this is a nice README it's hidden in this directory, could be of great help for users if documented in Edge docs instead of here https://wiki.polygon.technology/docs/edge/get-started/set-up-ibft-locally, wdyt?

File: Makefile
Comment (ZeljkoBenovic) [reply Makefile]: This local env var determines which consensus mechanism is to be used. It acts as a kind of a switch statement.   
The `docker-compose` file passes it to the `Dockerfile` as a second argument, which is used by the `entrypoint` script to create genesis for a specific consensus.
It must be set before the `init` container spins up, as it needs to know which consensus is being used.

The alternative to this approach is to have separate files for each consensus.    
We would than need to have dedicated `docker-compose`, `Dockerfile` and `entrypoint` script for each consensus.

IMHO, this is much simpler approach, as we're using the same files and the users can easily make consensus selection just by setting `EDGE_CONSENSUS` env var.

But if you disagree with this, we can easily split this logic.

File: docker/README.md
Comment (ZeljkoBenovic) [reply docker/README.md]: Great idea, and the users would definitely benefit from having something similar to this `README` in the public docs.    
We should probably do that, once `feature/v3-parity` is merged into develop and we have a release including this feature.    

I don't think that we have any information regarding `polybft` in our public docs yet.

File: consensus/polybft/polybft.go
Comment (Stefan-Ethernal) [inline consensus/polybft/polybft.go]: Nitpick: I would consider slightly changing this error message. Also similar should be done for `NativeTokenContract`. https://github.com/0xPolygon/polygon-edge/blob/df4e5dae55b2f32c6d7080d43f0c469b3248907d/consensus/polybft/polybft.go#L143

```suggestion
			return fmt.Errorf("ChildValidatorSet SC initialization failed: %w", err)
```

File: docker/local/polygon-edge.sh
Comment (Stefan-Ethernal) [inline docker/local/polygon-edge.sh:34]: Workflow has slightly changed with #1005...

Now as a prerequisite for  running a `genesis` command is `manifest` command. Manifest command is polybft specific and it creates json with validator public information as well as rootchain configuration (in case we are running polybft in bridge mode).

For additional information, you can either check `scripts/cluster` (https://github.com/0xPolygon/polygon-edge/blob/feature/v3-parity/scripts/cluster#L22) script and [README](https://github.com/0xPolygon/polygon-edge/blob/feature/v3-parity/consensus/polybft/README.md) or ping me. :slightly_smiling_face:

File: command/genesis/utils.go
Comment (Stefan-Ethernal) [inline command/genesis/utils.go]:
```suggestion
// ReadValidatorsByRegexp tries to find validator secrets when local secrets manager is used.
```

File: command/genesis/utils.go
Comment (Stefan-Ethernal) [inline command/genesis/utils.go]: Hm not sure is it safe to assume this will always be the case?
Especially now, when we have make possible that validator secrets are located on the same hierarchy levels, which means that directory names don't need to be unique.

File: command/genesis/utils.go
Comment (Stefan-Ethernal) [inline command/genesis/utils.go]: Just thinking out loud: is it a requirement to have validator secrets placed within different root folder(s)? Appreciate the effort, but not sure is it reasonable to make logic more complex to support this use case (especially if it is not an essential one)?

File: command/genesis/utils.go
Comment (Stefan-Ethernal) [inline command/genesis/utils.go]: Two things:
- `^.*\D-\d{1}$` with this regexp, I haven't managed to capture e.g.test-chain, so my question is should we even match non-digit here, since seems the result is the same as running this regexp `^.*-\d{1}$`? Can you give a counter-example?
- consider putting this regexp as const?

**Review by Stefan-Ethernal — COMMENTED** [review]
Tnx for effort. Leaving some comments and doubts to consider.

File: consensus/polybft/polybft.go
Comment (ZeljkoBenovic) [reply consensus/polybft/polybft.go]: Reworked the docker scripts, so that these code changes are not needed. 
Reverted all the code changes, so now only docker related files are changed.

File: Makefile
Comment (ZeljkoBenovic) [reply Makefile]: @vcastellm updated readme file with the instruction on how to run with `make` or directly with `docker-compose`.
Kept the same logic, as we're using existing docker files, but just switching the consensus with `EDGE_CONSENSUS` env var. This keeps the scripts simple and lean.     
Also added in some customization options for the chain.

File: docker/local/polygon-edge.sh
Comment (ZeljkoBenovic) [reply docker/local/polygon-edge.sh:34]: Reworked scripts to include the latest changes to the workflow.
It works as intended now.

**Review by Stefan-Ethernal — APPROVED** [review]
LGTM :rocket:

File: Makefile
Comment (vcastellm) [reply Makefile]: My proposal is to remove all the `run-local` stuff out from Makefile to a `docker` script in `scripts`, the fact that it's driven by env variables it's not related to using Makefile, it can be also improved by adding an option to the `cluster` script for running under docker, `scripts/cluster polybft --docker` should be the UI

**Review by vcastellm — DISMISSED** [review]
See https://github.com/0xPolygon/polygon-edge/pull/1051#discussion_r1056125288

**Review by vcastellm — COMMENTED** [review]
I don't want to block this for just the scripts thing, LGTM if we agree as a team that Makefile is the place

File: Makefile
Comment (ZeljkoBenovic) [reply Makefile]: @vcastellm I see now that we already have dedicated `scripts` folder. Edited the `cluster` script to accommodate `ibft` and `polybft` docker environments.     
Let me know if this works for you and the rest of the team.

Will change the PR description, but there are `README.md` files in both `docker` and `scripts` folder with some explanation how to use the script.

**Review by vcastellm — APPROVED** [review]
LGTM

**Pull request merged**
